### ISSUE TITLE
Clean code: Adopt project coding guidelines

### ISSUE THREAD

**Issue by douglasism** [microsoft/CCF#917]
**Description**
The project should adopt coding guidelines to improve consistency, efficiency, and maintainability across the code base. Problems like #777 could be resolved and avoided if shared guidelines were followed.  My recommendation is to adopt https://github.com/isocpp/CppCoreGuidelines as a baseline, and then add project exceptions where needed.

Additionally, using the Microsoft Guidelines Support Library https://github.com/Microsoft/GSL is also recommended. It includes functions and types that help projects follow the core guidelines.

**Comment by eddyashton** [comment]
I'm a big fan of coding guidelines, and the CppCoreGuidelines in particular.

I think we can (and do) _use_ guidelines (providing a reference when making a suggestion in a PR? Great!), but I'm less keen on _adopting_ guidelines (claiming our code follows this and all future changes must? Hard!). I'm particularly wary of adopting any policy that we can't automatically enforce, and I don't believe there are any tools that lint for most of the recommendations in these guidelines. I haven't looked at this for a while though - if you can recommend any tools for our CI pipeline to improve our code quality, we're very interested.

There's also a lot of work to make the existing codebase adhere to these guidelines (from low-hanging fruit like [consistent naming](https://github.com/microsoft/CCF/issues/707) to thorny problems like [auditing our exceptions](https://github.com/microsoft/CCF/issues/777)) - I don't think we'd gain much from this general compatibility work, versus fixing the specific deviations where they cause a problem.

**Comment by douglasism** [comment]
Hi Eddy, I'm not recommending an immediate, complete push for perfect adherence to the guidelines. They are a common goal that will move the code in the same direction. The guidelines themselves characterize my thoughts on this well:

> Consider these rules ideals for new code, opportunities to exploit when working on older code, and try to approximate these ideals as closely as feasible.  [In.aims: Aims](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#inaims-aims)

This section is a good initial read too: [In.0: Don't panic!](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#in0-dont-panic)

**Comment by douglasism** [comment]
And by "adopt" I meant that the guidelines, whichever set is used, should be explicitly stated (or linked to) in the Contributing section of https://github.com/microsoft/CCF/blob/master/README.md.

**Comment by achamayou** [comment]
@douglasism I have answered #777, and I do not believe that the lack of reference to a specific set of C++ guidelines in the README is to blame, nor that it would help. It's also not the case that all code in the repo can use C++ exceptions for error handling, nor that all code should handle errors in the exact same way. Because of the interaction with 3rd party components and language runtimes, a context-based approach is necessary.

If you have looked through the CI definitions and CMake files, you will notice that we make extensive use of clang tooling for static and dynamic analysis (formatting, diagnostics, scanner and sanitisers).

We welcome PRs that enable more of these checks and address the issues they uncover!

**Comment by douglasism** [comment]
Hi @achamayou,  Oh, I did not mean to imply that including a reference to C++ guidelines would be a magic wand. If only it were that simple. Improving the code base will require that the project contributors follow the guidelines (as possible). Sharing a common, way-of-our-people set of coding standards helps both the codebase and the team. For example, shared guidelines help frame code quality discussions during pre-merge code reviews. They also help the code have similar approaches and patterns, allowing contributors to move efficiently through the codebase.

**The Most Important Point**
Good coding guidelines are not about tabs-vs-spaces or where to put braces. Many of the most valuable and compelling guidelines in http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#main will not be detected by today’s linters or static analyzers (there’s promise in AI for this though!).

(Regarding #777: the main point of that issue is that error/exception handling in the codebase is incomplete and inconsistent. There does not seem to be a common approach or pattern, even at the class or module level.)

**Comment by achamayou** [comment]
@douglasism I'm wary of referencing such a large document, which we largely don't comply with today, and don't necessarily aim to comply with in any kind of well-defined time horizon. I much prefer documentation that reflects the actual truth, and I'd rather spend the effort to extend our use of automatic checks, which aren't subject to interpretation and yield clear results (the proof we comply is that the CI passes!).

I'm particularly supicious about the stuff that's not enforceable through tooling, because it tends to be fairly general opinions that don't apply usefully in every context. The fact that C++ guideline sets have proliferated rather than converged over time suggests to me that they aren't getting any better in this respect.

In short, I worry this is going to cause more discussion than it will solve, and will therefore not turn out to be a good investment for the project. With all that said, if there's consensus among the rest of team, I'm happy to be overruled on this topic :)

**Comment by douglasism** [comment]
I am working on one more valiant attempt at making my case for explicit, shared guidelines. I should be able to share them by next week. Please be patient and keep this open.

**Comment by ashamis** [comment]
Hi @douglasism, we have been adding some of these guidelines into our CI verification as has been appropriate. I am going to close this ticket as at this point there is nothing actionable, however, we will continue adding checks to our CI.

**Comment by achamayou** [comment]
@douglasism we now have CodeQL running with both extended security and code quality checks: https://github.com/microsoft/CCF/security/code-scanning?page=1&query=tool%3ACodeQL+is%3Aopen+autofilter%3Atrue++

I'm much more comfortable (and quite keen in fact) targeting this as a set of guidelines, and committing to a clean pass, but we'll have a team discussion first.
